### PR TITLE
Fix discount issue on checkout in VAT tax scenario #1336

### PR DESCRIPTION
**Pull request by SteveIreland**
This will hopefully resolve the reported issue.  I was unable to recreate it after these fixes were applied, but we know there might be a difference in configuration that triggers it.

### CODE REVIEW

File: application/libraries/Sale_lib.php [hunk 1]
```diff
@@ -329,27 +329,56 @@ public function get_totals()
...
+			error_log(">>> extended_discounted_amount-".$extended_discounted_amount);
+			$total = bcadd($total, $extended_discounted_amount);
+
+			if($this->CI->config->item('tax_included'))
 			{
-				$tax_exclusive_subtotal = bcadd($tax_exclusive_subtotal, $this->get_item_total_tax_exclusive($item['item_id'], $item['quantity'], $item['price'], $item['discount'], TRUE));
+				$subtotal = bcadd($subtotal, $this->get_extended_total_tax_exclusive($item['item_id'], $extended_discounted_amount, $item['quantity'], $item['price'], $item['discount']));
+				error_log(">>> subtotal-".$subtotal);
+
+//				$tax_exclusive_subtotal = bcadd($tax_exclusive_subtotal, $this->get_item_total_tax_exclusive($item['item_id'], $item['quantity'], $item['price'], $item['discount'], FALSE));
```
Comment (jekkos) [inline application/libraries/Sale_lib.php]: I'd remove these comments in case we don't need it anymore.. Also think the log statements should be on debug level instead of error

File: application/libraries/Sale_lib.php [hunk 2]
```diff
@@ -358,28 +387,34 @@ public function get_totals()
...
 		$totals['payment_total'] = $payment_total;
+		error_log(">>> payment_total-".$totals['payment_total']);
 
 		$amount_due = bcsub($total, $payment_total);
 		$totals['amount_due'] = $amount_due;
+		error_log(">>> amount_due-".$totals['amount_due']);
 
 		$cash_amount_due = bcsub($cash_total, $payment_total);
 		$totals['cash_amount_due'] = $cash_amount_due;
+		error_log(">>> cash_amount_due-".$totals['cash_amount_due']);
```
Comment (jekkos) [inline application/libraries/Sale_lib.php]: Actually these debugging statement belong in a testcase that describes and verifies expected amounts for certain situations

Comment (SteveIreland) [inline application/libraries/Sale_lib.php]: Ack ... My excuse is that it was late at night and I was exhausted.  I forgot the cleanup step again.   Sorry.  I will take care of this soon.

**Comment by odiea** [comment]
I guess I don't understand Vat but these numbers for a normal sale don't look right to me.
2017-06-29 07_11_40-wedidit _ powered by ospos 3 1 0

**Comment by odiea** [comment]
Sorry my ignorance for the previous post.

File: application/controllers/Sales.php [hunk 3]
```diff
@@ -468,24 +468,26 @@ public function complete()
 		$data["invoice_number"] = $this->sale_lib->get_invoice_number();
 		$data["quote_number"] = $this->sale_lib->get_quote_number();
 		$customer_info = $this->_load_customer_data($customer_id, $data);
-
+		if($customer_info != null)
```
Comment (daN4cat) [inline application/controllers/Sales.php]: please use NULL

File: application/controllers/Sales.php [hunk 4]
```diff
@@ -468,24 +468,26 @@ public function complete()
...
 
 		// Returns 'subtotal', 'total', 'cash_total', 'payment_total', 'amount_due', 'cash_amount_due', 'payments_cover_total'
 		$totals = $this->sale_lib->get_totals();
-		$data['subtotal'] = $totals['discounted_subtotal'];
-		$data['cash_total'] = $totals['cash_total'];
-		$data['cash_amount_due'] = $totals['cash_amount_due'];
-		$data['non_cash_total'] =$totals['total'];
-		$data['non_cash_amount_due'] =$totals['amount_due'];
+		$data['subtotal'] = $totals['subtotal'];
+		$data['total'] =$totals['total'];
```
Comment (daN4cat) [inline application/controllers/Sales.php]: there is a missing space across all the change of this line as it's a copy & paste story

File: application/libraries/Sale_lib.php [hunk 5]
```diff
@@ -329,27 +329,56 @@ public function get_totals()
 
 		$totals = array();
 
+		$prediscount_subtotal = 0;
 		$subtotal = 0;
+		$total = 0;
+		$total_discount = 0;
 		$discounted_subtotal = 0;
-		$tax_exclusive_subtotal = 0;
+//		$tax_exclusive_subtotal = 0;
```
Comment (daN4cat) [inline application/libraries/Sale_lib.php]: is // required?

File: application/libraries/Sale_lib.php [hunk 6]
```diff
@@ -329,27 +329,56 @@ public function get_totals()
...
 		foreach($this->get_cart() as $item)
 		{
-			$subtotal = bcadd($subtotal, $this->get_item_total($item['quantity'], $item['price'], $item['discount'], FALSE));
-			$discounted_subtotal = bcadd($discounted_subtotal, $this->get_item_total($item['quantity'], $item['price'], $item['discount'], TRUE));
-			if($this->CI->config->config['tax_included'])
+			$discount_amount = $this->get_item_discount($item['quantity'], $item['price'], $item['discount']);
+			error_log(">>> discount_amount-".$discount_amount);
```
Comment (daN4cat) [inline application/libraries/Sale_lib.php]: please remove all the error_log

File: application/models/Sale.php [hunk 7]
```diff
@@ -1169,6 +1169,46 @@ public function get_dinner_table($sale_id)
 	}
 
 	/*
+ * get the quote_number for the selected sale
```
Comment (daN4cat) [inline application/models/Sale.php]: indentation issues

File: application/views/sales/invoice.php [hunk 8]
```diff
@@ -163,7 +163,7 @@ function(response)
             <tr>
                 <td colspan="3" class="blank"> </td>
                 <td colspan="2" class="total-line"><textarea rows="5" cols="6"><?php echo $splitpayment[0]; ?></textarea></td>
-                <td class="total-value"><textarea rows="5" cols="6" id="paid"><?php echo to_currency( $payment['payment_amount'] * -1 ); ?></textarea></td>
+                <td class="total-value"><textarea rows="5" cols="6" id="paid"><?php echo to_currency( $payment['payment_amount'] ); ?></textarea></td>
```
Comment (daN4cat) [inline application/views/sales/invoice.php]: please revert this change as it's not agreed

File: application/views/sales/invoice.php [hunk 9]
```diff
@@ -183,7 +183,7 @@ function(response)
         <tr>
             <td colspan="3" class="blank"> </td>
             <td colspan="2" class="total-line"> <textarea rows="5" cols="6"><?php echo $this->lang->line($amount_change >= 0 ? ($only_sale_check ? 'sales_check_balance' : 'sales_change_due') : 'sales_amount_due') ; ?></textarea></td>
-            <td class="total-value"><textarea rows="5" cols="6" id="change"><?php echo to_currency($amount_change); ?></textarea></td>
+            <td class="total-value"><textarea rows="5" cols="6" id="change"><?php echo to_currency(abs($amount_change)); ?></textarea></td>
```
Comment (daN4cat) [inline application/views/sales/invoice.php]: same here not agreed

Comment (SteveIreland) [inline application/views/sales/invoice.php]: Well you are just in a disagreeable mood today.

Comment (daN4cat) [inline application/views/sales/invoice.php]: Yep, I know I can come across like that but on the other hand I have really little time and I focus on reviewing stuff as quick as possible and comments can be really straight to the point.

**Comment by daN4cat** [comment]
merged
